cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java

```diff
@@ -244,40 +246,41 @@ public void run(DatasetContext context) throws Exception {
 
         // This block has been added so that completed workflow runs can be logged to the workflow dataset
+        WorkflowId workflowId = new WorkflowId(id.getParent(), id.getProgram());
         if (id.getType() == ProgramType.WORKFLOW && runStatus == ProgramRunStatus.COMPLETED) {
-          recordCompletedWorkflow(metaStore, getWorkflowDataset(context),
-                                  Id.Workflow.from(id.getParent().toId(), id.getProgram()), pid);
+          recordCompletedWorkflow(metaStore, getWorkflowDataset(context), workflowId, pid);
         }
         // todo: delete old history data
       }
     });
   }
 
   private void recordCompletedWorkflow(AppMetadataStore metaStore, WorkflowDataset workflowDataset,
-                                       Id.Workflow workflowId, String runId) {
-    RunRecordMeta runRecord = metaStore.getRun(workflowId.toEntityId(), runId);
+                                       WorkflowId workflowId, String runId) {
+    RunRecordMeta runRecord = metaStore.getRun(workflowId, runId);
     if (runRecord == null) {
       return;
     }
-    Id.Application app = workflowId.getApplication();
-    ApplicationSpecification appSpec = getApplicationSpec(metaStore, workflowId.getApplication().toEntityId());
-    if (appSpec == null || appSpec.getWorkflows() == null || appSpec.getWorkflows().get(workflowId.getId()) == null) {
+    ApplicationId app = workflowId.getParent();
+    ApplicationSpecification appSpec = getApplicationSpec(metaStore, workflowId.getParent());
+    if (appSpec == null || appSpec.getWorkflows() == null
+      || appSpec.getWorkflows().get(workflowId.getProgram()) == null) {
       LOG.warn("Missing ApplicationSpecification for {}, " +
                  "potentially caused by application removal right after stopping workflow {}", app, workflowId);
       return;
     }
 
     boolean workFlowNodeFailed = false;
-    WorkflowSpecification workflowSpec = appSpec.getWorkflows().get(workflowId.getId());
+    WorkflowSpecification workflowSpec = appSpec.getWorkflows().get(workflowId.getProgram());
     Map<String, WorkflowNode> nodeIdMap = workflowSpec.getNodeIdMap();
     final List<WorkflowDataset.ProgramRun> programRunsList = new ArrayList<>();
     for (Map.Entry<String, String> entry : runRecord.getProperties().entrySet()) {
       if (!("workflowToken".equals(entry.getKey()) || "runtimeArgs".equals(entry.getKey())
         || "workflowNodeState".equals(entry.getKey()))) {
         WorkflowActionNode workflowNode = (WorkflowActionNode) nodeIdMap.get(entry.getKey());
         ProgramType programType = ProgramType.valueOfSchedulableType(workflowNode.getProgram().getProgramType());
-        Id.Program innerProgram = Id.Program.from(app.getNamespaceId(), app.getId(), programType, entry.getKey());
-        RunRecordMeta innerProgramRun = metaStore.getRun(innerProgram.toEntityId(), entry.getValue());
+        ProgramId innerProgram = new ProgramId(app, programType, entry.getKey());
```

Comment:
does this compile? I don't see a 3 argument constructor in ProgramId. Just use ``app.program(programType, entry.getKey())``